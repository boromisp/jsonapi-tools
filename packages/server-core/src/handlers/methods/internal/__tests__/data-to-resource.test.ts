import 'jest';
import dataToResource from '../data-to-resource';

const schema = {
  type: 'test-type',
  relationships: {
    rel: {
      type: 'related-type'
    },
    'other-rel': {
      type: 'related-type'
    }
  }
};

const itemId = '1';

describe('Test building an IResourceObject', () => {
  test('Let nulls pass through', () => {
    expect(dataToResource(schema, null)).toBeNull();
  });

  test('Convert `empty` IResrouceData to IResourceObject', () => {
    expect(dataToResource(schema, {
      id: itemId,
      type: schema.type
    })).toEqual({
      id: itemId,
      type: schema.type,
      links: { self: `/${schema.type}/${itemId}` }
    });
  });

  test('Add attributes', () => {
    const date = new Date();

    expect(dataToResource(schema, {
      id: itemId,
      type: schema.type,
      some: 'attribute',
      'a-date': date,
      'an-array-of-stuff': [1, '2', { n: '3' }]
    })).toEqual({
      id: itemId,
      type: schema.type,
      links: { self: `/${schema.type}/${itemId}` },
      attributes: {
        some: 'attribute',
        'a-date': date,
        'an-array-of-stuff': [1, '2', { n: '3' }]
      }
    });
  });

  test('Add relationships', () => {
    expect(dataToResource(schema, {
      id: itemId,
      type: schema.type,
      rel: '5',
      'other-rel': null
    })).toEqual({
      id: itemId,
      type: schema.type,
      links: { self: `/${schema.type}/${itemId}` },
      relationships: {
        rel: {
          data: {
            type: 'related-type',
            id: '5'
          },
          links: {
            self:  `/${schema.type}/${itemId}/relationships/rel`,
            related: `/${schema.type}/${itemId}/rel`
          }
        },
        'other-rel': {
          data: null,
          links: {
            self:  `/${schema.type}/${itemId}/relationships/other-rel`,
            related: `/${schema.type}/${itemId}/other-rel`
          }
        }
      }
    });
  });

  test('Use custom `links` generation', () => {
    expect(dataToResource({
      type: 'test-type',
      links({ id }: { id: string; }) {
        return { self: { href: `/v2/test-type/${id}`, meta: { version: 'v2' } } };
      }
    }, { id: itemId, type: schema.type })).toEqual({
      id: itemId,
      type: schema.type,
      links: { self: { href: `/v2/test-type/${itemId}`, meta: { version: 'v2' } } }
    });
  });
});
