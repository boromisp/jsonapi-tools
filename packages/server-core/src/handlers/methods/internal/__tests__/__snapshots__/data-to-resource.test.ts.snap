// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test building an IResourceObject Add attributes 1`] = `
Object {
  "attributes": Object {
    "a-date": 2009-02-13T23:31:30.123Z,
    "an-array-of-stuff": Array [
      1,
      "2",
      Object {
        "n": "3",
      },
    ],
    "some": "attribute",
  },
  "id": "1",
  "links": Object {
    "self": "/test-type/1",
  },
  "type": "test-type",
}
`;

exports[`Test building an IResourceObject Add relationships 1`] = `
Object {
  "id": "1",
  "links": Object {
    "self": "/test-type/1",
  },
  "relationships": Object {
    "other-rel": Object {
      "data": null,
      "links": Object {
        "related": "/test-type/1/other-rel",
        "self": "/test-type/1/relationships/other-rel",
      },
    },
    "rel": Object {
      "data": Object {
        "id": "5",
        "type": "related-type",
      },
      "links": Object {
        "related": "/test-type/1/rel",
        "self": "/test-type/1/relationships/rel",
      },
    },
  },
  "type": "test-type",
}
`;

exports[`Test building an IResourceObject Convert \`empty\` IResrouceData to IResourceObject 1`] = `
Object {
  "id": "1",
  "links": Object {
    "self": "/test-type/1",
  },
  "type": "test-type",
}
`;

exports[`Test building an IResourceObject Let nulls pass through 1`] = `null`;

exports[`Test building an IResourceObject Use custom \`links\` generation 1`] = `
Object {
  "id": "1",
  "links": Object {
    "self": Object {
      "href": "/v2/test-type/1",
      "meta": Object {
        "version": "v2",
      },
    },
  },
  "type": "test-type",
}
`;
